homeassistant/components/sia/hub.py

```diff
@@ -0,0 +1,90 @@
+"""The sia hub."""
+import logging
+
+from pysiaalarm.aio import CommunicationsProtocol, SIAAccount, SIAClient, SIAEvent
+
+from homeassistant.const import CONF_PORT, CONF_PROTOCOL, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, EventOrigin, HomeAssistant
+from homeassistant.helpers import device_registry as dr
+
+from .const import (
+    CONF_ACCOUNT,
+    CONF_ACCOUNTS,
+    CONF_ENCRYPTION_KEY,
+    CONF_IGNORE_TIMESTAMPS,
+    DOMAIN,
+    SIA_EVENT,
+)
+
+DEFAULT_TIMEBAND = (80, 40)
+IGNORED_TIMEBAND = (3600, 1800)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class SIAHub:
+    """Class for SIA Hubs."""
+
+    def __init__(
+        self, hass: HomeAssistant, hub_config: dict, entry_id: str, title: str
+    ):
+        """Create the SIAHub."""
+        self._hass = hass
+        self._port = int(hub_config[CONF_PORT])
+        self.entry_id = entry_id
+        self._title = title
+        self._accounts = hub_config[CONF_ACCOUNTS]
+
+        self._remove_shutdown_listener = None
+        self.sia_accounts = [
+            SIAAccount(
+                account_id=a[CONF_ACCOUNT],
+                key=a.get(CONF_ENCRYPTION_KEY),
+                allowed_timeband=IGNORED_TIMEBAND
+                if a[CONF_IGNORE_TIMESTAMPS]
+                else DEFAULT_TIMEBAND,
+            )
+            for a in self._accounts
+        ]
+        self.sia_client = SIAClient(
+            host="",
+            port=self._port,
+            accounts=self.sia_accounts,
+            function=self.async_create_and_fire_event,
+            protocol=CommunicationsProtocol(hub_config[CONF_PROTOCOL]),
+        )
+
+    async def async_setup_hub(self):
+        """Add a device to the device_registry, register shutdown listener, load reactions."""
+        device_registry = await dr.async_get_registry(self._hass)
+        port = self._port
+        for acc in self._accounts:
+            account = acc[CONF_ACCOUNT]
+            device_registry.async_get_or_create(
+                config_entry_id=self.entry_id,
+                identifiers={(DOMAIN, port, account)},
+                name=f"{port} - {account}",
+            )
+        self._remove_shutdown_listener = self._hass.bus.async_listen(
+            EVENT_HOMEASSISTANT_STOP, self.async_shutdown
+        )
+
+    async def async_shutdown(self, _: Event = None):
+        """Shutdown the SIA server."""
+        if self._remove_shutdown_listener:
+            self._remove_shutdown_listener()
+        await self.sia_client.stop()
+
+    async def async_create_and_fire_event(self, event: SIAEvent):
+        """Create a event on HA's bus, with the data from the SIAEvent."""
+        _LOGGER.debug(
+            "Adding event to bus for code %s for account %s and port %s",
+            event.code,
+            event.account,
+            self._port,
+        )
+        self._hass.bus.async_fire(
```

Comment:
If the event is being consumed by an entity, we should not fire it since we end up with two events (state_changed) and the event itself

Reply:
This is actually the expected behaviour, I also created a note on that in the function message. But there are many SIA Codes that might come in, including for things not built into the HA integration, currently only 22 out of ~320 are handled all in the alarm_control_panel, there will be binary sensors in a follow-on PR which will handle a couple more, but every code that comes in might be relevant for a user, so with this event they can then add their own automation on top of those events (or the events that are also handled by the integration).